source/types.ts

```diff
@@ -1,274 +1,121 @@
-import CacheableLookup from 'cacheable-lookup';
-import {Except, Merge} from 'type-fest';
-import {GotReturn} from './create';
-import {GotError, HTTPError, MaxRedirectsError, ParseError, TimeoutError, RequestError} from './errors';
-import {Hooks} from './known-hook-events';
-import {URLOptions} from './utils/options-to-url';
-
-export type GeneralError = Error | GotError | HTTPError | MaxRedirectsError | ParseError;
-
-export type Method =
-	| 'GET'
-	| 'POST'
-	| 'PUT'
-	| 'PATCH'
-	| 'HEAD'
-	| 'DELETE'
-	| 'OPTIONS'
-	| 'TRACE'
-	| 'get'
-	| 'post'
-	| 'put'
-	| 'patch'
-	| 'head'
-	| 'delete'
-	| 'options'
-	| 'trace';
-
-export type ResponseType = 'json' | 'buffer' | 'text';
-
-export interface Response<BodyType = unknown> extends IncomingMessageWithTimings {
-	body: BodyType;
-	statusCode: number;
-
-	/**
-	The remote IP address.
-
-	Note: Not available when the response is cached. This is hopefully a temporary limitation, see [lukechilds/cacheable-request#86](https://github.com/lukechilds/cacheable-request/issues/86).
-	*/
-	ip: string;
-
-	fromCache?: boolean;
-	isFromCache?: boolean;
-	req?: http.ClientRequest;
-	requestUrl: string;
-	retryCount: number;
-	timings: Timings;
-	redirectUrls: string[];
-	request: {
-		options: NormalizedOptions;
-	};
-	url: string;
-}
-
-// TODO: The `ResponseLike` type should be properly fixed instead:
-// https://github.com/sindresorhus/got/pull/827/files#r323633794
-export interface ResponseObject extends Partial<ResponseLike> {
-	socket: {
-		remoteAddress: string;
-	};
-}
-
-export interface RetryObject {
-	attemptCount: number;
-	retryOptions: Required<RetryOptions>;
-	error: TimeoutError | RequestError;
-	computedValue: number;
+import {CancelError} from 'p-cancelable';
+import {
+	// Request & Response
+	CancelableRequest,
+	Response,
+
+	// Options
+	Options,
+	NormalizedOptions,
+	Defaults as DefaultOptions,
+	PaginationOptions,
+
+	// Errors
+	ParseError,
+	RequestError,
+	CacheError,
+	ReadError,
+	HTTPError,
+	MaxRedirectsError,
+	TimeoutError
+} from './as-promise';
+import Request from './core';
+
+export interface InstanceDefaults {
+	options: DefaultOptions;
+	handlers: HandlerFunction[];
+	mutableDefaults: boolean;
+	_rawHandlers?: HandlerFunction[];
 }
 
-export type RetryFunction = (retryObject: RetryObject) => number;
-
+export type GotReturn = Request | CancelableRequest;
 export type HandlerFunction = <T extends GotReturn>(options: NormalizedOptions, next: (options: NormalizedOptions) => T) => T | Promise<T>;
 
-export interface DefaultRetryOptions {
-	limit: number;
-	methods: Method[];
-	statusCodes: number[];
-	errorCodes: string[];
-	calculateDelay: RetryFunction;
-	maxRetryAfter?: number;
-}
-
-export interface RetryOptions extends Partial<DefaultRetryOptions> {
-	retries?: number;
-}
-
-export type RequestFunction = typeof http.request;
-
-export interface AgentByProtocol {
-	http?: http.Agent;
-	https?: https.Agent;
-}
```

Comment:
Breaking change: `options.agent` can no longer be an instance of `http.Agent`. Instead, you need to pass an object like:

```
{
	http: new http.Agent(...),
	https: new https.Agent(...),
	http2: new http2wrapper.Agent(...)`
}
```

Reply:
Can you mention all the breaking changes in the PR description?

Reply:
Will do! :D